### DISCUSSION TITLE
Issues in mesoscale det restart

### DISCUSSION THREAD

**Discussion by PerryShafran-NOAA** [NOAA-EMC/EVS#601] [Issues]
Hi, everyone!

I need some assistance in diagnosing an issue here.  With help from @MarcelCaron-NOAA, I installed restart in the mesoscale stats jobs for NAM and RAP.  I noticed that when I do a restart job, the final stat file is much smaller than it would be for a full job.  I found out why; not all the stat files from the stmp working directory makes it over to the restart directory.  

Compare the following two directories:

working directory: `/lfs/h2/emc/stmp/perry.shafran/evs_test/prod/tmp/jevs_mesoscale_rap_grid2obs_stats_00.159901212.cbqs01/grid2obs/METplus_output/raob/point_stat/rap.20241030`

restart directory:
`/lfs/h2/emc/vpppg/noscrub/perry.shafran/evs/v2.0/stats/mesoscale/atmos.20241030/rap/grid2obs/restart/c07/METplus_output/raob/point_stat/rap.20241030`

These are the two directories after the job was killed after 7 minutes.  Note that the working directory has 1397 stat files in it, while the restart directory has only 290 files in it.  

The codebase can be found here:  `/lfs/h2/emc/vpppg/noscrub/perry.shafran/EVS_mesoscale_v2/EVS`

Relevant job file is here:  `/lfs/h2/emc/vpppg/noscrub/perry.shafran/EVS_mesoscale_v2/EVS/dev/drivers/scripts/stats/mesoscale/jevs_mesoscale_rap_grid2obs_stats.sh`

This job file is usually run with a `-v vhr=07` setting.  

The latest job log is here:  `/lfs/h2/emc/vpppg/noscrub/perry.shafran/EVS_mesoscale_v2/EVS/dev/drivers/scripts/stats/mesoscale/jevs_mesoscale_rap_grid2obs_stats_00.o159901212`

I had made changes in the ush/mesoscale directory, and thus, checking out some of those scripts might be helpful as I might have missed or deleted a line somewhere that I shouldn't have.  I'm not 100% familiar with these scripts as I wasn't the original developer, but I'm figuring stuff out little by little.  Nevertheless, I feel stuck here and if anyone could offer some assistance/guidance, that would be great.

I think there may be similar issues in the NAM run, but I'll run that now to offer an additional data point.  NAM and RAP use the same ush scripts, though they have different ex-scripts.  

Thanks, all!

Perry

**Comment by malloryprow** [comment]
@PerryShafran-NOAA Do you have the log from the initial interrupted run?

**Comment by PerryShafran-NOAA** [comment]
@malloryprow The log file listed above *is* the interrupted run.  In this case I ran only the interrupted run so I could compare what's in the restart directory vs the working directory.  They differ significantly, which is likely why we have the smaller final file when we do the restart run.

**Reply by malloryprow** [reply]
Ohhhh, okay I see I see.

**Comment by malloryprow** [comment]
I think I found it. Look at Looking at /lfs/h2/emc/stmp/perry.shafran/evs_test/prod/tmp/jevs_mesoscale_rap_grid2obs_stats_00.159901212.cbqs01/grid2obs/stats/METplus_job_scripts/generate/job130 to follow what I am saying.

The problem is that mesoscale runs with these different `FHR_GROUP` values. It is either `SHORT` or `FULL`. This sets `FHR_END`, `FHR_INCR`,  and `MIN_IHOUR` (`FHR_START` gets calculated from this and `VHOUR` and `FHR_INCR`) depending on the value for `FHR_GROUP`. METplus gets run with these different settings for `FHR_GROUP=SHORT`, and then gets run again for `FHR_GROUP=FULL`. So for /lfs/h2/emc/stmp/perry.shafran/evs_test/prod/tmp/jevs_mesoscale_rap_grid2obs_stats_00.159901212.cbqs01/grid2obs/METplus_output/raob/point_stat/rap.20241030/point_stat_rap_ak_HGT_OBS_*L_20241030_060000V.stat we see 27 files from both METplus runs.

When mesoscale_util gets called to run copy_data_to_restart, it is being run outsides of the `FHR_GROUP` loop. Making the input settings `FHR_START`, `FHR_END`, and `FHR_INCR` are coming from when `FHR_GROUP=FULL` thus missing the files generated when `FHR_GROUP=SHORT`.

If you put the copy_to_restart call within in the loop, you should be good.

**Reply by PerryShafran-NOAA** [reply]
Great that you found where the issue is!  Can you show me precisely where to make these changes?  I'm not sure I follow the code that well.

**Reply by malloryprow** [reply]
I didn't look at the code at all to debug, just the output from your run's DATA, so I don't know where and in what scripts this is occuring.

**Reply by PerryShafran-NOAA** [reply]
OK, thanks!  Maybe @MarcelCaron-NOAA can assist?

**Reply by malloryprow** [reply]
If I remember right, there are a handful of python scripts related to making the job files so it might be somewhere in those?

**Reply by PerryShafran-NOAA** [reply]
Yeah, I kind of know which files to look at (mesoscale_stats_grid2obs_create_job_script.py) - but I'm not sure I know enough about the python to figure out where the looping for FHR_GROUP is and to make these changes.  I have a feeling that the changes could be somewhat substantial.

**Reply by PerryShafran-NOAA** [reply]
Also, FWIW, I'm seeing the same thing in cam (for example, /lfs/h2/emc/stmp/emc.vpppg/evs_beta5_test/prod/tmp/jevs_cam_hrrr_grid2obs_stats.159906653.cbqs01/grid2obs/stats/METplus_job_scripts/generate/job8).  Thus, this issue *may* be in CAM as well.

**Reply by malloryprow** [reply]
I think you just need to look where the loop is being written and move the copy_data_to_restart call in there. It shouldn't be too bad other than a copy paste.

**Reply by malloryprow** [reply]
Noted about cam. @MarcelCaron-NOAA can you take a look and confirm cam is having this as well. We can add it to the Fixes and Additions document if needed.

**Reply by MarcelCaron-NOAA** [reply]
@PerryShafran-NOAA yes, mesoscale_stats_grid2obs_create_job_script.py is probably the file to fix.  When writing the job cards, some commands are added to a regular list of commands—which are just appended to the end of the card—and some are added to an "iterative" command list, which are written inside the for loops on the card.  The restart command probably is added to the regular list when it should be added to the "iterative" list.

**Reply by MarcelCaron-NOAA** [reply]
@malloryprow Yup let me take a look!

**Reply by PerryShafran-NOAA** [reply]
Oh! I think that may be the missing piece.  Thus, I'd need to change the following lines (441-456) from this:

```
             job_cmd_list.append(
               f'python -c '
               + '\"import mesoscale_util as cutil; cutil.copy_data_to_restart('
               + '\\\"${DATA}\\\", \\\"${RESTART_DIR}\\\", '
               + 'verif_case=\\\"${VERIF_CASE}\\\", '
               + 'verif_type=\\\"${VERIF_TYPE}\\\", '
               + 'vx_mask=\\\"${NEST}\\\", '
               + 'met_tool=\\\"point_stat\\\", '
               + 'vdate=\\\"${VDATE}\\\", '
               + 'vhour=\\\"${VHOUR}\\\", '
               + 'fhr_start=\\\"${FHR_START}\\\", '
               + 'fhr_end=\\\"${FHR_END}\\\", '
               + 'fhr_incr=\\\"${FHR_INCR}\\\", '
               + 'model=\\\"${MODELNAME}\\\", '
               + 'var_name=\\\"${VAR_NAME}\\\"'
               + ')\"'
            )   
```

to this:

```
            job_cmd_list_iterative.append(
               f'python -c '
               + '\"import mesoscale_util as cutil; cutil.copy_data_to_restart('
               + '\\\"${DATA}\\\", \\\"${RESTART_DIR}\\\", '
               + 'verif_case=\\\"${VERIF_CASE}\\\", '
               + 'verif_type=\\\"${VERIF_TYPE}\\\", '
               + 'vx_mask=\\\"${NEST}\\\", '
               + 'met_tool=\\\"point_stat\\\", '
               + 'vdate=\\\"${VDATE}\\\", '
               + 'vhour=\\\"${VHOUR}\\\", '
               + 'fhr_start=\\\"${FHR_START}\\\", '
               + 'fhr_end=\\\"${FHR_END}\\\", '
               + 'fhr_incr=\\\"${FHR_INCR}\\\", '
               + 'model=\\\"${MODELNAME}\\\", '
               + 'var_name=\\\"${VAR_NAME}\\\"'
               + ')\"'
            )   

```

I'm going to try that.

**Reply by MarcelCaron-NOAA** [reply]
Right, I think that's the fix.  There _may_ be some downstream issues (may not want to put that command inside all loops, just the FCST_GROUP loop) but we can test and check.

**Reply by MarcelCaron-NOAA** [reply]
@malloryprow @PerryShafran-NOAA also yes, this exists in cam too.  Can we add another cam-det fix item for this issue?

**Reply by PerryShafran-NOAA** [reply]
I'm only changing here in the one place and nowhere else.  Let's see what happens.

**Reply by AliciaBentley-NOAA** [reply]
@MarcelCaron-NOAA I added it to the Fixes and Additions document under cam (det.)! I'll also link to this discussion.

**Reply by PerryShafran-NOAA** [reply]
Hmm, that didn't work.  The copy_data_to_restart command is still outside the FHR_GROUP, as in /lfs/h2/emc/stmp/perry.shafran/evs_test/prod/tmp/jevs_mesoscale_rap_grid2obs_stats_00.159914542.cbqs01/grid2obs/stats/METplus_job_scripts/generate/job130.  Maybe that was the wrong place to change to iterative.  Let me look at this further.

**Reply by MarcelCaron-NOAA** [reply]
@AliciaBentley-NOAA Thanks!

@PerryShafran-NOAA - Looks like every "copy_data_to_restart" command under "generate" and "reformat" groups need to be added to the "job_cmd_list_iterative" list of commands (2 instances for "reformat" and ~10 instances for "generate").  Probably inconsequential for "gather*".

**Reply by PerryShafran-NOAA** [reply]
Yes - I added it in a second place.  It's improving:  now 938 stat files were copied to the restart directory compared to 1446 in the working directory.  I didn't add it to reformat.  I'll add it in all reformat and generate locations and try one more time.

**Reply by PerryShafran-NOAA** [reply]
@malloryprow @MarcelCaron-NOAA Even as I added every instance of copy_data_to_restart to all the iterative commands under both generate and reformat, the number of files are still not equal (938 for restart, 1446 for the working directory).  

Could it simply be now that the job ended after the data was created in METplus but before the copy to restart was able to be done?  Perhaps.  But if data is missing from the restart directory, presumably the code now would see what's missing and go ahead and create those stat files, in theory.  

Let me try a restarted run and see what I get, now that I have made the iterative changes.  Presumably, every single file that exists in restart is copied to the new restart working directory, and anything else not there is done in the restart job.  Does this make sense?

**Comment by malloryprow** [comment]
Yes that is possible. From looking yesterday if you see files that are not fully completed to the restart directory but that job is marked as completed then there is a problem.

**Reply by PerryShafran-NOAA** [reply]
@malloryprow @MarcelCaron-NOAA It's still a problem despite marking all the calls to copy_data_to_restart iterative in generate and reformat.  

In the latest run, here are the working and restart directories for the interrupted job:
working directory (1445 files): `/lfs/h2/emc/stmp/perry.shafran/evs_test/prod/tmp/jevs_mesoscale_rap_grid2obs_stats_00.159997973.cbqs01/grid2obs/METplus_output/raob/point_stat/rap.20241031`

restart directory (938 files): `/lfs/h2/emc/vpppg/noscrub/perry.shafran/evs/v2.0/stats/mesoscale/atmos.20241031/rap/grid2obs_interrupt/restart/c07/METplus_output/raob/point_stat/rap.20241031`

Job log: `/lfs/h2/emc/vpppg/noscrub/perry.shafran/EVS_mesoscale_v2/EVS/dev/drivers/scripts/stats/mesoscale/jevs_mesoscale_rap_grid2obs_stats_00.o159997973`

And thus, the final stat files are different sizes. See: 

```
38 (clogin09) /lfs/h2/emc/vpppg/noscrub/perry.shafran/evs/v2.0/stats/mesoscale/rap.20241031 > ls -l
total 187220
-rw-r--r-- 1 perry.shafran vpppg 60557909 Nov  1 13:12 evs.stats.rap.atmos.grid2obs.v20241031.stat
-rw-r--r-- 1 perry.shafran vpppg 65573103 Nov  1 12:14 evs.stats.rap.atmos.grid2obs.v20241031.stat_cron
-rw-r--r-- 1 perry.shafran vpppg 65573103 Nov  1 12:36 evs.stats.rap.atmos.grid2obs.v20241031.stat_full
39 (clogin09) /lfs/h2/emc/vpppg/noscrub/perry.shafran/evs/v2.0/stats/mesoscale/rap.20241031
```

Obviously there are still some files that are not being copied over from the working directory to the restart directory.  I am not entirely sure why that would be.  More eyes on the output/code would be helpful for me in figuring this out.

I haven't even talked about metar, but there are issues there too:

working directory (7329 files):
`/lfs/h2/emc/stmp/perry.shafran/evs_test/prod/tmp/jevs_mesoscale_rap_grid2obs_stats_00.159997973.cbqs01/grid2obs/METplus_output/metar/point_stat/rap.20241031`

restart directory (7034 files):
`/lfs/h2/emc/vpppg/noscrub/perry.shafran/evs/v2.0/stats/mesoscale/atmos.20241031/rap/grid2obs_interrupt/restart/c07/METplus_output/metar/point_stat/rap.20241031`

**Reply by PerryShafran-NOAA** [reply]
As another data point, here are the stat files for the restart job:

working directory (1910 files): `/lfs/h2/emc/stmp/perry.shafran/evs_test/prod/tmp/jevs_mesoscale_rap_grid2obs_stats_00.159999546.cbqs01/grid2obs/METplus_output/raob/point_stat/rap.20241031`

restart directory (1874 files):
`/lfs/h2/emc/vpppg/noscrub/perry.shafran/evs/v2.0/stats/mesoscale/atmos.20241031/rap/grid2obs/restart/c07/METplus_output/raob/point_stat/rap.20241031`

Job log: `/lfs/h2/emc/vpppg/noscrub/perry.shafran/EVS_mesoscale_v2/EVS/dev/drivers/scripts/stats/mesoscale/jevs_mesoscale_rap_grid2obs_stats_00.o159999546`

And the full job (no restart).

working directory (1946 files): `/lfs/h2/emc/stmp/perry.shafran/evs_test/prod/tmp/jevs_mesoscale_rap_grid2obs_stats_00.159996319.cbqs01/grid2obs/METplus_output/raob/point_stat/rap.20241031`

restart directory (1874 files):
`/lfs/h2/emc/vpppg/noscrub/perry.shafran/evs/v2.0/stats/mesoscale/atmos.20241031/rap/grid2obs_full/restart/c07/METplus_output/raob/point_stat/rap.20241031`

Job log: `/lfs/h2/emc/vpppg/noscrub/perry.shafran/EVS_mesoscale_v2/EVS/dev/drivers/scripts/stats/mesoscale/jevs_mesoscale_rap_grid2obs_stats_00.o159996319`

It should be noted that the final stat files are created from the stat files in the working directory,  They differ in the final restart job vs. the full job (no restart).  In the full job, the restart directory is still not the same as the working directory.  The key here is making sure that all stat files in the working directory are copied over to the restart directory in com, and that's still not happening in either case, but it harms the restart job more to be sure.

**Reply by malloryprow** [reply]
I think the fhr loop in copy_to_restart for point_stat isn't inclusive of the last forecast hour and it is missing the file for the last forecast hour.  I'm looking at line 571 in mesoscale_util.py. 

What tipped me off to this was 
`ls /lfs/h2/emc/vpppg/noscrub/perry.shafran/evs/v2.0/stats/mesoscale/atmos.20241031/rap/grid2obs_interrupt/restart/c07/METplus_output/raob/point_stat/rap.20241031/point_stat_rap_ak_HGT_OBS*060000V.stat | wc -l` is 26

`ls /lfs/h2/emc/stmp/perry.shafran/evs_test/prod/tmp/jevs_mesoscale_rap_grid2obs_stats_00.159997973.cbqs01/grid2obs/METplus_output/raob/point_stat/rap.20241031/point_stat_rap_ak_HGT_OBS*_060000V.stat | wc -l` is 27.

If you list them out you'll see f051 is missing from /lfs/h2/emc/vpppg/noscrub/perry.shafran/evs/v2.0/stats/mesoscale/atmos.20241031/rap/grid2obs_interrupt/restart/c07/METplus_output/raob/point_stat/rap.20241031/point_stat_rap_ak_HGT_OBS*060000V.stat files.

**Reply by PerryShafran-NOAA** [reply]
Yes, I did notice in the final stat file that there were no records for 51000 lead time, as there were when I did a full job with no restart, and found that weird.  I wonder if that might be the final piece of the puzzle, to figure out why the f051 data seems to be excluded.  Let me see if I can figure it out.

**Reply by PerryShafran-NOAA** [reply]
There seem to be 36 files in the working directory, so that makes up part of the difference, but not all.

However, all 36 files for the 450000 lead time made it to the restart directory.  

Let me check for every restart time how many files there are.

**Reply by PerryShafran-NOAA** [reply]
For the f000 time, there are 72 files for the working directory, only 36 for the restart.  No f000 files for 00Z or 12Z valid time made it to the restart directory.  The same for f001 time and for f006 time (and presumably for f002-f005 as well).

**Comment by PerryShafran-NOAA** [comment]
@malloryprow Returning to this issue here, as now I have put mpmd into the system, and have worked, with Marcel's help, to incorporate restart back into the mpmd code.  However, this issue remains:  with restart, the final stat file is still smaller than it is without restart.  

```
6 (clogin04) /lfs/h2/emc/vpppg/noscrub/perry.shafran/evs/v2.0/stats/mesoscale/nam.20250113 > ls -l
total 80240
-rw-r--r-- 1 perry.shafran vpppg 40263331 Jan 14 16:41 evs.stats.nam.atmos.grid2obs.v20250113.stat
-rw-r--r-- 1 perry.shafran vpppg 41900283 Jan 14 15:18 evs.stats.nam.atmos.grid2obs.v20250113.stat_save
```

The _save file is the output file without restart (and matches what is found in emc.vpppg for the date). The other one was created when the script was submitted with a wallclock for 15 minutes.  Then the wallclock was reset back to 1 hr and submitted again without clearing the stats directory.  As you can see, it's significantly smaller.  

I do not know why stats files do not make it to the restart directory.  In many cases, the working directory had more files in the restart directory than the noscrub restart directory.  It seems that when the code ends, there are some small stat files having been created but didn't get copied because the end of the code occurred after file was generated but before the copy.  

I'd like some more help looking into this issue.  

Perry

**Reply by malloryprow** [reply]
What are your COMOUT and various DATA directories and logs?

**Reply by PerryShafran-NOAA** [reply]
small stats files: /lfs/h2/emc/vpppg/noscrub/perry.shafran/evs/v2.0/stats/mesoscale/atmos.20250113/nam/grid2obs
large stats files: /lfs/h2/emc/vpppg/noscrub/perry.shafran/evs/v2.0/stats/mesoscale/nam.20250113

.o file for full run:  /lfs/h2/emc/vpppg/noscrub/perry.shafran/EVS_mesoscale_restart/EVS/dev/drivers/scripts/stats/mesoscale/jevs_mesoscale_nam_grid2obs_stats_00.o176988955
.o file for interrupted run:  /lfs/h2/emc/vpppg/noscrub/perry.shafran/EVS_mesoscale_restart/EVS/dev/drivers/scripts/stats/mesoscale/jevs_mesoscale_nam_grid2obs_stats_00.o176990058
.o file for restart run:  /lfs/h2/emc/vpppg/noscrub/perry.shafran/EVS_mesoscale_restart/EVS/dev/drivers/scripts/stats/mesoscale/jevs_mesoscale_nam_grid2obs_stats_00.o176994416

working directory for full run:  /lfs/h2/emc/stmp/perry.shafran/evs_test/prod/tmp/jevs_mesoscale_nam_grid2obs_stats_00.176988955.cbqs01
working directory for interrupted run: /lfs/h2/emc/stmp/perry.shafran/evs_test/prod/tmp/jevs_mesoscale_nam_grid2obs_stats_00.176990058.cbqs01
working directory for restart run: /lfs/h2/emc/stmp/perry.shafran/evs_test/prod/tmp/jevs_mesoscale_nam_grid2obs_stats_00.176994416.cbqs01

**Reply by malloryprow** [reply]
Are you working from the branch [feature/mesoscale_restart](https://github.com/PerryShafran-NOAA/EVS/tree/feature/mesoscale_restart)?

**Reply by PerryShafran-NOAA** [reply]
Yes.

**Reply by malloryprow** [reply]
In [ush/mesoscale/mesoscale_util.py]( https://github.com/PerryShafran-NOAA/EVS/blob/feature/mesoscale_restart/ush/mesoscale/mesoscale_util.py) on lines 447, 466, 486, 510, 554, 576, 621, 650, change it to `for fhr in np.arange(int(fhr_start), int(fhr_end) + int(fhr_incr), int(fhr_incr)):`

**Reply by malloryprow** [reply]
See if that does anything

**Reply by PerryShafran-NOAA** [reply]
Working on it!

**Reply by PerryShafran-NOAA** [reply]
Good news!  After making these changes, it turns out that the full file after restart is now the exact same size at it was for the no-restart case!  I guess this works.  How does this change work?

**Reply by malloryprow** [reply]
numpy arange isn't inclusive so it wasn't including the last forecast hour.

**Reply by PerryShafran-NOAA** [reply]
Great, thanks!  I appreciate your help.  

Let's keep this open until the PR that I eventually submit is merged in.

**Reply by malloryprow** [reply]
Sounds like a plan; you're welcome!

**Comment by PerryShafran-NOAA** [comment]
Closing the discussion.
